// Copyright 2020 The XLS Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto2";

package xls;

message TypeProto {
  enum TypeEnum {
    INVALID = 0;
    BITS = 1;
    TUPLE = 2;
    ARRAY = 3;
    TOKEN = 4;
  }
  optional TypeEnum type_enum = 1;

  // For BITS types, this is the bit width.
  optional int64 bit_count = 2;

  // For TUPLE types, the type of the elements.
  repeated TypeProto tuple_elements = 3;

  // For ARRAY types, the number and type of the elements.
  optional int64 array_size = 4;
  optional TypeProto array_element = 5;
}

message FunctionTypeProto {
  repeated TypeProto parameters = 1;
  optional TypeProto return_type = 2;
}
